botbot_plugins/plugins/vote.py

```diff
@@ -0,0 +1,394 @@
+            return unicode(e)
+
+    # This makes the vote symbol (+, -, \) required, since otherwise we can't
+    # tell if this is meant to be a vote or not.
+    @listens_to_all(vote_regex_template.format(vote_modifier=""))
+    @_depends_on_votestarted(throw_error=False)
+    def implicit_vote(self, line, vote, option):
+        try:
+            parsed_vote = self._parse_vote(line.user, (vote, option))
+        except InvalidOptionError:
+            return  # Don't complain, will generate too many false positives
+        changes_made = self._add_vote(parsed_vote)
+        if changes_made:
+            self.store("changes_since_countvote", json.dumps(True))
+
+    @listens_to_command("abstain")
+    @_depends_on_votestarted()
+    def abstain(self, line, args):
+        self.store("changes_since_countvote", json.dumps(True))
+        self._add_vote(self._get_abstain_vote(line.user))
+
+    @listens_to_command("cancelvotes")
+    @_depends_on_votestarted()
+    def cancelvotes(self, line, args):
+        self.store("changes_since_countvote", json.dumps(True))
+        votes = json.loads(self.retrieve("votes"))
+        votes = self._remove_all_votes(line.user, votes)
+        self.store("votes", json.dumps(votes))
+
+    def _parse_vote(self, voter, (vote, option)):
+        boolean_shortform = self.config["boolean_shortform"]
+        if vote == "\\":
+            if (option == boolean_shortform or
+                    option == ABSTAIN_NAME):
+                return self._get_abstain_vote(voter)
+            else:
+                raise InvalidOptionError(ERROR_MESSAGES["invalid_abstain"]
+                                         .format(shortform=boolean_shortform))
+        if option.strip() == "":
+            option = boolean_shortform
+
+        options = self._get_valid_options()
+        longest_matching_option =\
+            self._find_longest_matching_option(option, options)
+        if longest_matching_option is None:
+            raise InvalidOptionError(ERROR_MESSAGES["invalid_option"]
+                                     .format(option=option))
+
+        return Vote(
+            voter=voter,
+            section=(longest_matching_option.strip()
+                     if longest_matching_option != boolean_shortform
+                     else DEFAULT_VOTE_OPTION),
+            vote=VOTE_SHORTCUT_TO_NAME[vote]
+        )
+
+    def _find_longest_matching_option(self, user_option, valid_options):
+        user_option = user_option.strip().expandtabs(1)
+
+        def matches_user_option(option):
+            option = option.strip()
+            # In startswith, adding the space ensures that "oreo"
+            # is not treated as matching the option "o".
+            return (user_option == option or
+                    user_option.startswith(option + " "))
+
+        matching_options = filter(matches_user_option, valid_options)
+        return (max(matching_options, key=len)
+                if matching_options else None)
```

Comment:
I am assuming this returns a string. Any reason for the return type discrepancy? It might cause a lot of bugs if not handled correctly.

Reply:
Do you mean, sometimes returning a string and sometimes returning `None`? I could change it to empty string, but I'm not clear on this.

Reply:
Yup. That's what I meant. An empty string would be better.